pam/integration-tests/ssh_test.go

```diff
@@ -0,0 +1,557 @@
+
+	return sb.Buffer.Write(p)
+}
+
+func (sb *safeBuffer) ReadFrom(r io.Reader) (n int64, err error) {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+
+	return sb.Buffer.ReadFrom(r)
+}
+
+func (sb *safeBuffer) String() string {
+	sb.mu.RLock()
+	defer sb.mu.RUnlock()
+
+	return sb.Buffer.String()
+}
+
+func startSSHd(t *testing.T, hostKey, forcedCommand string, env []string, daemonize bool) string {
+	t.Helper()
+
+	// We use this to easily find a free port we can use, without going random
+	server := httptest.NewServer(http.HandlerFunc(nil))
+	url, err := url.Parse(server.URL)
+	require.NoError(t, err, "Setup: Impossible to find a valid port to use")
+	sshdPort := url.Port()
+	server.Close()
+
+	sshd, sshdPidFile, sshdLogFile := sshdCommand(t, sshdPort, hostKey, forcedCommand, env, daemonize)
+	sshdStderr := safeBuffer{}
+	sshd.Stderr = &sshdStderr
+	if testing.Verbose() {
+		sshd.Stdout = os.Stdout
+		sshd.Stderr = os.Stderr
+	}
+
+	t.Log("Launching sshd with", sshd.Env, sshd.Args)
+	err = sshd.Start()
+	require.NoError(t, err, "Setup: Impossible to start sshd")
+
+	t.Cleanup(func() {
+		if testing.Verbose() || !t.Failed() {
+			return
+		}
+		sshdLog := filepath.Join(t.TempDir(), "sshd.log")
+		require.NoError(t, os.WriteFile(sshdLog, []byte(sshdStderr.String()), 0600),
+			"TearDown: Saving sshd log")
+		saveArtifactsForDebug(t, []string{sshdLog})
+	})
+
+	t.Cleanup(func() {
+		if sshd.Process == nil {
+			return
+		}
+
+		sshdExited := make(chan *os.ProcessState)
+		go func() {
+			processState, err := sshd.Process.Wait()
+			require.NoError(t, err, "TearDown: Waiting SSHd failed")
+			sshdExited <- processState
+		}()
+
+		t.Log("Waiting for sshd to be terminated")
+		select {
+		case <-time.After(sleepDuration(30 * time.Second)):
+			require.NoError(t, sshd.Process.Kill(), "TearDown: Killing SSHd failed")
+			if !testing.Verbose() {
+				t.Logf("SSHd stopped (killed)\n ##### STDERR #####\n %s \n ##### END #####",
+					sshdStderr.String())
+			}
+			t.Fatal("SSHd didn't finish in time!")
+		case state := <-sshdExited:
+			if !testing.Verbose() {
+				t.Logf("SSHd stopped (%s)\n ##### STDERR #####\n %s \n ##### END #####",
+					state, sshdStderr.String())
+			}
+			expectedExitCode := 255
+			if daemonize {
+				expectedExitCode = 0
+			}
+			require.Equal(t, expectedExitCode, state.ExitCode(), "TearDown: SSHd exited with %s", state)
+		}
+	})
+
+	if !daemonize {
+		// Sadly we can't wait for SSHd to be ready using net.Dial, since that will make sshd
+		// (when in debug mode) not to accept further connections from the actual test, but we
+		// can assume we're good.
+		t.Logf("SSHd started with pid %d and listening on port %s", sshd.Process.Pid, sshdPort)
+		return sshdPort
+	}
+
+	t.Cleanup(func() {
+		if !t.Failed() && !testutils.IsVerbose() {
+			return
+		}
+		contents, err := os.ReadFile(sshdLogFile)
+		require.NoError(t, err, "TearDown: Reading SSHd log failed")
+		t.Logf(" ##### LOG FILE #####\n %s \n ##### END #####", contents)
```

Comment:
Well this only happens on failures or when using `-v`, personally I prefer to have everything inlined, an in case going though the artifacts for details. Also this contains our module logging.

Also given that the artifacts path isn't know at this level, it would imply refactoring quite a bit `saveArtifactsForDebug` so that we can point out where the file would be... So, honestly for now I'd leave in this way, as I didn't notice it being too long for being parsed.

Reply:
Ok, let's leave it for now and revisit if we notice in practice that it's making our life harder.